ddtrace/profiling/_periodic.py

```diff
@@ -32,24 +33,23 @@ def __init__(self, interval, target, name=None, on_shutdown=None):
         self.quit = threading.Event()
         self.daemon = True
 
-    def start(self):
-        """Start the thread."""
-        super(PeriodicThread, self).start()
-        PERIODIC_THREAD_IDS.add(self.ident)
```

Comment:
This is where the race occurred. The thread is started in the line above so there is the possibility that it runs to completion, calling `PERIODIC_THREAD_IDS.remove(self.ident)` before it has been added.

Reply:
Why you don't just invert those lines?

It'd actually be better and safer to have it added before we start the thread for real since the profiler will be able to identify sooner the thread as being a profiler periodic thread (and ignore it).

Reply:
`self.ident` is not defined until `_bootstrap()` further down the call stack from `super(PeriodicThread, self).start()`.

https://github.com/python/cpython/blob/fa0a66e62d087765dbc5c1b89d6149a23ecfb0a6/Lib/threading.py#L933-L935

Reply:
(The very first thing I tried was to invert them 😛)

Reply:
Damn…

Reply:
Do you think it's worth it to override `_bootstrap_inner` in order set it as soon as possible? We do already do it for gevent.

Reply:
No, I think it's fine to keep this simple. Worst case you'll see a tiny sample that won't be identified as being from the profiler, but since it's technically impossible to make it perfect anyway, let's not make things complicated.